### PR TITLE
BUG: prevent coercion to datetime64[ns] when a Series is initialized with both tz-naive and tz-aware

### LINKED ISSUE #16406: BUG: construction of tz-naive and tz-aware should not coerce
```
In [1]: Series([Timestamp('20130101'), Timestamp('20130101', tz='US/Eastern')])
Out[1]: 
0   2012-12-31 19:00:00-05:00
1   2013-01-01 00:00:00-05:00
dtype: datetime64[ns, US/Eastern]
```

should instead be

```
In [2]: Series([Timestamp('20130101'), Timestamp('20130101', tz='US/Eastern')], dtype=object)
Out[2]: 
0          2013-01-01 00:00:00
1    2013-01-01 00:00:00-05:00
dtype: object
```

### PR DESCRIPTION
**Pull request by jamestran201**
- [x] closes #16406 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/_libs/tslib.pyx [hunk 1]
```diff
@@ -1252,12 +1252,18 @@ def datetime_to_datetime64(ndarray[object] values):
...
     result = np.empty(n, dtype='M8[ns]')
     iresult = result.view('i8')
+    tz_naive_or_tz_aware = None
     for i in range(n):
         val = values[i]
         if _checknull_with_nat(val):
             iresult[i] = NPY_NAT
         elif PyDateTime_Check(val):
             if val.tzinfo is not None:
+                if tz_naive_or_tz_aware is None:
```
Comment (jbrockmendel) [inline pandas/_libs/tslib.pyx]: What is this accomplishing that ‘inferred_tz’ isn’t?

**Comment by jreback** [comment]
so where you are fixing is not the correct location, its a bit after the actual error, rather try here:
https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/src/inference.pyx#L459

a mixed datetime64 & datetime64 that is tz-aware are in the same bucket, rather this needs disambiguation, and ultimately should be 'mixed'.

Note we have *no* tests for this routine, so that is the first thing we need. Pls add to 
``pandas/tests/dtypes/test_inference.py``

**Comment by tmnhat2001** [comment]
my bad, should have checked out that routine

**Comment by tmnhat2001** [comment]
np.datetime64 objects don't have timezone info right? I just want to make sure not to leave it out.

**Comment by jreback** [comment]
no, no.datetime64 don’t support time zone (they print with a local time zone which is extra confusing)

File: doc/source/whatsnew/v0.22.0.txt [hunk 2]
```diff
@@ -178,5 +178,5 @@ Other
 ^^^^^
 
 - Improved error message when attempting to use a Python keyword as an identifier in a numexpr query (:issue:`18221`)
--
+- Bug in :func:`inference.pyx.infer_to_datetimelike_array` that coerces tz-naive and tz-aware values in the same array to datetime64[ns] (:issue:`16406`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: give  a user perspective here (IOW don't reference internal functions).

File: pandas/core/series.py [hunk 3]
```diff
@@ -3238,7 +3240,8 @@ def _try_cast(arr, take_fast_path):
         else:
             subarr = maybe_convert_platform(data)
 
-        subarr = maybe_cast_to_datetime(subarr, dtype)
+        if try_cast_to_datetime:
```
Comment (jreback) [inline pandas/core/series.py]: why is this changed?

File: pandas/tests/dtypes/test_inference.py [hunk 4]
```diff
@@ -655,6 +655,69 @@ def test_infer_dtype_period(self):
...
+    )
+    def test_infer_datetimelike_array_timedelta(self, data):
+        assert lib.infer_datetimelike_array(data) == "timedelta"
+
+    def test_infer_datetimelike_array_date(self):
+        arr = [date(2017, 6, 12), date(2017, 3, 11)]
+        assert lib.infer_datetimelike_array(arr) == "date"
+
+    @pytest.mark.parametrize(
+        "data",
```
Comment (jreback) [inline pandas/tests/dtypes/test_inference.py]: can you add strings, ints, float, and a python object 

e.g. 
```
class Foo():
    pass;
Foo()
```
should all return mixed (as this is a date infered)

File: pandas/tests/series/test_timeseries.py [hunk 5]
```diff
@@ -949,3 +949,8 @@ def test_get_level_values_box(self):
         index = MultiIndex(levels=levels, labels=labels)
 
         assert isinstance(index.get_level_values(0)[0], Timestamp)
+
+    def test_tz_naive_and_tz_aware_mix(self):
+        s = Series([Timestamp('20130101'),
```
Comment (jreback) [inline pandas/tests/series/test_timeseries.py]: add the issue number as a comment

do a full comparision with assert_series_equal, IOW construct the expected

move this test to 
test_constructors.py

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/series.py (hunk 3 above)
Comment (tmnhat2001) [reply pandas/core/series.py]: sorry about that, i was trying out something similar to infer_datetimelike_array in that function before you told me to try in inference.pyx. I'll change it back to the original version.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@tmnhat2001 nice patch! keep em coming!

**Comment by tmnhat2001** [comment]
thanks
